File: src/main/java/org/cubexmc/metro/train/TrainMovementTask.java
package org.cubexmc.metro.train;

import java.util.Map;

import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.entity.Minecart;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.HandlerList;
import org.bukkit.event.Listener;
import org.bukkit.event.vehicle.VehicleMoveEvent;
import org.bukkit.util.Vector;
import org.cubexmc.metro.Metro;
import org.cubexmc.metro.event.TrainEnterStopEvent;
import org.cubexmc.metro.manager.LanguageManager;
import org.cubexmc.metro.manager.LineManager;
import org.cubexmc.metro.manager.StopManager;
import org.cubexmc.metro.model.Line;
import org.cubexmc.metro.model.PriceRule;
import org.cubexmc.metro.model.Stop;
import org.cubexmc.metro.service.TicketService;
import org.cubexmc.metro.util.SchedulerUtil;

/**
 * Controls one event-driven train ride from a stop to the next stop.
 */
public class TrainMovementTask implements Listener {

    public enum TrainState {
        STOPPED_AT_STATION,
        MOVING_IN_STATION,
        MOVING_BETWEEN_STATIONS
    }

    private final TrainSession session;
    private final TrainStateMachine stateMachine;
    private final TrainScheduler trainScheduler;
    private final TrainPhysicsController physicsController;
    private final TrainEventPublisher eventPublisher;
    private final TrainScoreboardController scoreboardController;
    private final TrainMovementAssistController movementAssistController;

    public TrainMovementTask(Metro plugin, Minecart minecart, Player passenger, String lineId, String fromStopId) {
        this(plugin, minecart, passenger, lineId, fromStopId, TrainState.STOPPED_AT_STATION);
    }

    public TrainMovementTask(Metro plugin, Minecart minecart, Player passenger, String lineId, String fromStopId,
            TrainState initialState) {
        LineManager lineManager = plugin.getLineManager();
        Line line = lineManager.getLine(lineId);
        this.session = new TrainSession(plugin, minecart, passenger, line, fromStopId, initialState);
        this.stateMachine = new TrainStateMachine(session);
        this.trainScheduler = new TrainScheduler(plugin);
        this.physicsController = new TrainPhysicsController();
        this.eventPublisher = new TrainEventPublisher(session);
        this.scoreboardController = new TrainScoreboardController();
        this.movementAssistController = new TrainMovementAssistController(session, trainScheduler,
                physicsController, this::cancel, this::handlePassengerExit);

        if (line != null && session.getTargetStopId() != null) {
            scoreboardController.updateBasedOnState(session);
        }
    }

    public void cancel() {
        Minecart minecart = session.getMinecart();
        TrainTaskRegistry.unregister(minecart);
        movementAssistController.stop();
        trainScheduler.cancelAll();
        HandlerList.unregisterAll(this);
        session.debug("Task cancelled for passenger=" + session.safePassengerName()
                + ", currentStop=" + session.getCurrentStopId()
                + ", targetStop=" + session.getTargetStopId());
    }

    public void removeMinecartAndCancel() {
        Minecart minecart = session.getMinecart();
        if (minecart != null && !minecart.isDead()) {
            minecart.eject();
            minecart.remove();
        }
        cancel();
    }

    void removeMinecartAndCancelOnEntityScheduler() {
        Minecart minecart = session.getMinecart();
        if (minecart == null) {
            cancel();
            return;
        }
        SchedulerUtil.entityRun(session.getPlugin(), minecart, this::removeMinecartAndCancel, 0L, -1L);
    }

    public void transferMinecart(Minecart newCart) {
        Minecart previousCart = session.getMinecart();
        session.setMinecart(newCart);
        session.setTeleporting(false);
        TrainTaskRegistry.transfer(previousCart, newCart, this);
        Line line = session.getLine();
        if (line != null && session.getPlugin().getRouteRecorder() != null) {
            session.getPlugin().getRouteRecorder().transferCart(line.getId(), previousCart, newCart);
        }
        if (session.getState() == TrainState.MOVING_BETWEEN_STATIONS) {
            movementAssistController.start();
        }
        session.debug("Transferred movement task to new minecart UUID=" + newCart.getUniqueId());
    }

    public void setTeleporting(boolean teleporting) {
        session.setTeleporting(teleporting);
    }

    public boolean canUsePortal(String portalId) {
        Line line = session.getLine();
        return line != null && line.containsPortal(portalId);
    }

    public static TrainMovementTask getTaskFor(Minecart cart) {
        return TrainTaskRegistry.get(cart);
    }

    public static int shutdownActiveTasks() {
        return TrainTaskRegistry.shutdownActiveTasks();
    }

    public static int shutdownActiveTasks(Metro plugin, boolean folia) {
        return TrainTaskRegistry.shutdownActiveTasks(plugin, folia);
    }

    Object scheduleSessionTask(Runnable task, long delay, long period) {
        return trainScheduler.entityRun(session.getMinecart(), task, delay, period);
    }

    TrainSession getSession() {
        return session;
    }

    @EventHandler(priority = EventPriority.NORMAL)
    public void onTrainEnterStop(TrainEnterStopEvent event) {
        if (!event.getMinecart().equals(session.getMinecart())) {
            return;
        }

        if (!session.isTeleporting() && !session.isPassengerStillRiding()) {
            handlePassengerExit();
            return;
        }

        Stop enteredStop = event.getStop();
        if (enteredStop != null && session.getTargetStopId() != null
                && session.getTargetStopId().equals(enteredStop.getId())) {
            transitionToMovingInStation(enteredStop);
        }
    }

    @EventHandler(priority = EventPriority.HIGH)
    public void onVehicleMove(VehicleMoveEvent event) {
        Minecart minecart = session.getMinecart();
        if (!event.getVehicle().equals(minecart)) {
            return;
        }

        Line line = session.getLine();
        if (line != null) {
            session.getPlugin().getRouteRecorder().sample(line.getId(), minecart, event.getTo());
        }
        updateLastTravelDirection(event.getFrom(), event.getTo());

        if (session.getState() == TrainState.MOVING_BETWEEN_STATIONS) {
            Location from = event.getFrom();
            Location to = event.getTo();
            if (from.getWorld() != null && to.getWorld() != null && from.getWorld().equals(to.getWorld())) {
                double dx = to.getX() - from.getX();
                double dz = to.getZ() - from.getZ();
                double dist = Math.sqrt(dx * dx + dz * dz);
                if (dist > 0.001) {
                    session.addDistance(dist);
                }
            }
        }

        if (session.getState() != TrainState.MOVING_IN_STATION) {
            return;
        }

        Stop targetStop = session.getPlugin().getStopManager().getStop(session.getTargetStopId());
        if (targetStop == null || targetStop.getStopPointLocation() == null) {
            return;
        }

        Location currentLocation = minecart.getLocation();
        Location targetLocation = targetStop.getStopPointLocation();
        if (currentLocation.getWorld() == null || !currentLocation.getWorld().equals(targetLocation.getWorld())) {
            return;
        }

        double distance = currentLocation.distance(targetLocation);
        if (distance < 0.8) {
            transitionToStoppedAtStation(targetStop);
            return;
        }

        physicsController.applyApproachBraking(minecart, distance,
                session.getPlugin().getConfigFacade().getCartSpeed());
    }

    private void transitionToStoppedAtStation(Stop stop) {
        Minecart minecart = session.getMinecart();
        minecart.setVelocity(new Vector(0, 0, 0));
        minecart.setMaxSpeed(0);
        movementAssistController.stop();

        Location snapLocation = stop.getStopPointLocation().clone();
        snapLocation.setX(snapLocation.getBlockX() + 0.5);
        snapLocation.setZ(snapLocation.getBlockZ() + 0.5);
        if (session.getLine() != null) {
            session.getPlugin().getRouteRecorder().sample(session.getLine().getId(), minecart, snapLocation);
        }
        SchedulerUtil.teleportEntity(minecart, snapLocation);

        settleDistanceFare(stop);

        TrainState previousState = stateMachine.transitionTo(TrainState.STOPPED_AT_STATION, null);
        if (previousState == TrainState.MOVING_IN_STATION) {
            handleArrivalAtStation();
        }

        scoreboardController.updateBasedOnState(session);
    }

    private void settleDistanceFare(Stop stop) {
        Line line = session.getLine();
        if (line == null) return;

        PriceRule rule = line.getPriceRule();
        if (rule == null) return;

        double distance = session.getDistanceTraveled();
        if (distance <= 0 && rule.getMode() != PriceRule.PricingMode.INTERVAL) return;

        Player passenger = session.getPassenger();
        if (passenger == null || !passenger.isOnline()) return;

        double variablePrice = 0;

        if (rule.getMode() == PriceRule.PricingMode.DISTANCE) {
            variablePrice = distance * rule.getPerBlockRate();
        } else if (rule.getMode() == PriceRule.PricingMode.INTERVAL) {
            int intervals = session.getPlugin().getPriceService()
                    .countStopIntervals(line, session.getEntryStopId(), stop.getId());
            variablePrice = intervals * rule.getPerIntervalRate();
        } else {
            return;
        }

        if (variablePrice > 0) {
            TicketService.TicketChargeStatus status = session.getPlugin().getTicketService()
                    .chargePrice(passenger, line, variablePrice);
            if (status == TicketService.TicketChargeStatus.CHARGED) {
                passenger.sendMessage(session.getPlugin().getLanguageManager().getMessage("economy.paid_distance",
                        LanguageManager.put(LanguageManager.args(), "price",
                                session.getPlugin().getTicketService().format(variablePrice))));
            }
        }

        session.addDistance(-distance);
    }

    private void transitionToMovingInStation(Stop targetStop) {
        movementAssistController.stop();

        TrainState previousState = stateMachine.transitionTo(TrainState.MOVING_IN_STATION,
                "enteredStop=" + targetStop.getId());
        if (previousState == TrainState.MOVING_BETWEEN_STATIONS) {
            eventPublisher.publishEnteringStop(targetStop);
        }

        scoreboardController.updateBasedOnState(session);
    }

    private void transitionToMovingBetweenStations() {
        stateMachine.transitionTo(TrainState.MOVING_BETWEEN_STATIONS, null);

        Player passenger = session.getPassenger();
        if (passenger != null && passenger.isOnline()) {
            passenger.sendTitle("", "", 0, 0, 0);
        }
        scoreboardController.updateBasedOnState(session);
    }

    private void handleArrivalAtStation() {
        handleArrivalAtStation(false);
    }

    void startAtStation() {
        handleArrivalAtStation(true);
    }

    private void handleArrivalAtStation(boolean isNewlySpawned) {
        Line line = session.getLine();
        if (line == null) {
            cancel();
            return;
        }

        if (!isNewlySpawned) {
            session.setCurrentStopId(session.getTargetStopId());
        }

        Stop currentStop = session.getPlugin().getStopManager().getStop(session.getCurrentStopId());
        if (currentStop == null) {
            session.debug("Cancelling train because current stop is missing: " + session.getCurrentStopId());
            cancel();
            return;
        }

        if (line.getNextStopId(session.getCurrentStopId()) == null) {
            if (!handleTerminalStation()) {
                return;
            }
            session.setTargetStopId(null);
            eventPublisher.publishDockedAtStop(currentStop, true);
            return;
        }

        eventPublisher.publishDockedAtStop(currentStop, false);
        scoreboardController.updateBasedOnState(session);
        scheduleNextDeparture();
    }

    private void handleDeparture() {
        Line line = session.getLine();
        if (line == null) {
            cancel();
            return;
        }

        session.refreshTargetFromCurrentStop();
        if (session.getTargetStopId() == null) {
            return;
        }

        StopManager stopManager = session.getPlugin().getStopManager();
        Stop currentStop = stopManager.getStop(session.getCurrentStopId());
        Stop nextStop = stopManager.getStop(session.getTargetStopId());
        if (currentStop == null || nextStop == null || currentStop.getStopPointLocation() == null) {
            session.debug("Cancelling train because departure stops are incomplete: current="
                    + session.getCurrentStopId() + ", target=" + session.getTargetStopId());
            cancel();
            return;
        }

        session.getPlugin().getRouteRecorder().sample(line.getId(), session.getMinecart(),
                currentStop.getStopPointLocation());
        eventPublisher.publishDeparture(currentStop, nextStop);

        double maxSpeed = line.getMaxSpeed();
        if (maxSpeed == -1.0) {
            maxSpeed = session.getPlugin().getConfigFacade().getCartSpeed();
        }
        session.getMinecart().setMaxSpeed(maxSpeed);

        Vector launchDirection = physicsController.initMinecartVelocity(session.getMinecart(),
                currentStop.getLaunchYaw());
        if (launchDirection != null) {
            session.setLastTravelDirection(launchDirection);
        }

        transitionToMovingBetweenStations();
        movementAssistController.start();
    }

    private void scheduleNextDeparture() {
        long delay = session.getPlugin().getConfigFacade().getCartDepartureDelay();
        session.debug("Schedule departure in " + delay + " ticks for passenger="
                + session.safePassengerName() + ", currentStop=" + session.getCurrentStopId());
        trainScheduler.entityRun(session.getMinecart(), () -> {
            if (session.isPassengerStillRiding()) {
                handleDeparture();
            } else {
                handlePassengerExit();
            }
        }, delay, -1);
    }

    private boolean handleTerminalStation() {
        Player passenger = session.getPassenger();
        Line line = session.getLine();
        if (passenger == null || !passenger.isOnline() || line == null) {
            cancel();
            return false;
        }
        session.debug("Terminal station reached for passenger=" + session.safePassengerName()
                + ", stop=" + session.getCurrentStopId());

        org.cubexmc.metro.manager.RouteRecorder.FinishResult routeResult =
                session.getPlugin().getRouteRecorder().finishIfRecording(line.getId(), session.getMinecart());
        notifyRouteRecorder(routeResult);
        if (routeResult.status() == org.cubexmc.metro.manager.RouteRecorder.FinishResult.Status.TOO_FEW_POINTS) {
            session.getPlugin().getLogger().warning("[RouteRecorder] Route recording for line " + line.getId()
                    + " reached the terminal but only collected " + routeResult.pointCount() + " point(s).");
        }

        trainScheduler.entityRun(session.getMinecart(), () -> {
            Minecart minecart = session.getMinecart();
            if (minecart != null && !minecart.isDead()) {
                minecart.eject();
                session.getPlugin().getScoreboardManager().clearPlayerDisplay(passenger);

                trainScheduler.entityRun(minecart, () -> {
                    if (minecart != null && !minecart.isDead()) {
                        minecart.remove();
                    }
                    cancel();
                }, 40L, -1);
            } else {
                cancel();
            }
        }, 60L, -1);
        return true;
    }

    private void notifyRouteRecorder(org.cubexmc.metro.manager.RouteRecorder.FinishResult result) {
        if (result.recorderId() == null
                || result.status() == org.cubexmc.metro.manager.RouteRecorder.FinishResult.Status.NOT_RECORDING) {
            return;
        }
        Player recorder = Bukkit.getPlayer(result.recorderId());
        if (recorder == null || !recorder.isOnline()) {
            return;
        }

        Map<String, Object> args = LanguageManager.args();
        LanguageManager.put(args, "line_id", result.lineId());
        LanguageManager.put(args, "point_count", String.valueOf(result.pointCount()));

        String key = switch (result.status()) {
            case SAVED -> "line.record_saved";
            case TOO_FEW_POINTS -> "line.record_too_few";
            case FAILED -> "line.record_failed";
            case NOT_RECORDING -> null;
        };
        if (key != null) {
            recorder.sendMessage(session.getPlugin().getLanguageManager().getMessage(key, args));
        }
        if (result.status() == org.cubexmc.metro.manager.RouteRecorder.FinishResult.Status.TOO_FEW_POINTS) {
            recorder.sendMessage(session.getPlugin().getLanguageManager().getMessage("line.record_too_few_hint"));
        }
    }

    private void handlePassengerExit() {
        cancel();
    }

    public static void startTrainTask(Metro plugin, Minecart minecart, Player passenger, String lineId,
            String currentStopId) {
        TrainTaskStarter.start(plugin, minecart, passenger, lineId, currentStopId);
    }

    private void updateLastTravelDirection(Location from, Location to) {
        if (session.getState() == TrainState.STOPPED_AT_STATION || from == null || to == null) {
            return;
        }
        if (from.getWorld() == null || to.getWorld() == null || !from.getWorld().equals(to.getWorld())) {
            return;
        }
        Vector direction = to.toVector().subtract(from.toVector());
        if (direction.lengthSquared() < 0.0001) {
            return;
        }
        session.setLastTravelDirection(direction.normalize());
    }
}

File: src/main/java/org/cubexmc/metro/train/TrainMovementTask.kt
package org.cubexmc.metro.train

import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.entity.Minecart
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.HandlerList
import org.bukkit.event.Listener
import org.bukkit.event.vehicle.VehicleMoveEvent
import org.bukkit.util.Vector
import org.cubexmc.metro.Metro
import org.cubexmc.metro.event.TrainEnterStopEvent
import org.cubexmc.metro.manager.LanguageManager
import org.cubexmc.metro.model.PriceRule
import org.cubexmc.metro.model.Stop
import org.cubexmc.metro.service.TicketService
import org.cubexmc.metro.util.SchedulerUtil

/**
 * Controls one event-driven train ride from a stop to the next stop.
 */
class TrainMovementTask @JvmOverloads constructor(
    plugin: Metro,
    minecart: Minecart?,
    passenger: Player?,
    lineId: String?,
    fromStopId: String?,
    initialState: TrainState = TrainState.STOPPED_AT_STATION,
) : Listener {
    enum class TrainState {
        STOPPED_AT_STATION,
        MOVING_IN_STATION,
        MOVING_BETWEEN_STATIONS,
    }

    private val session: TrainSession
    private val stateMachine: TrainStateMachine
    private val trainScheduler: TrainScheduler
    private val physicsController: TrainPhysicsController
    private val eventPublisher: TrainEventPublisher
    private val scoreboardController: TrainScoreboardController
    private val movementAssistController: TrainMovementAssistController

    init {
        val line = plugin.lineManager.getLine(lineId)
        session = TrainSession(plugin, minecart, passenger, line, fromStopId, initialState)
        stateMachine = TrainStateMachine(session)
        trainScheduler = TrainScheduler(plugin)
        physicsController = TrainPhysicsController()
        eventPublisher = TrainEventPublisher(session)
        scoreboardController = TrainScoreboardController()
        movementAssistController =
            TrainMovementAssistController(
                session,
                trainScheduler,
                physicsController,
                Runnable { cancel() },
                Runnable { handlePassengerExit() },
            )

        if (line != null && session.targetStopId != null) {
            scoreboardController.updateBasedOnState(session)
        }
    }

    fun cancel() {
        val minecart = session.minecart
        TrainTaskRegistry.unregister(minecart)
        movementAssistController.stop()
        trainScheduler.cancelAll()
        HandlerList.unregisterAll(this)
        session.debug(
            "Task cancelled for passenger=${session.safePassengerName()}, currentStop=${session.currentStopId}, " +
                "targetStop=${session.targetStopId}",
        )
    }

    fun removeMinecartAndCancel() {
        val minecart = session.minecart
        if (minecart != null && !minecart.isDead) {
            minecart.eject()
            minecart.remove()
        }
        cancel()
    }

    fun removeMinecartAndCancelOnEntityScheduler() {
        val minecart = session.minecart
        if (minecart == null) {
            cancel()
            return
        }
        SchedulerUtil.entityRun(session.plugin, minecart, Runnable { removeMinecartAndCancel() }, 0L, -1L)
    }

    fun transferMinecart(newCart: Minecart) {
        val previousCart = session.minecart
        session.minecart = newCart
        session.isTeleporting = false
        TrainTaskRegistry.transfer(previousCart, newCart, this)
        val line = session.line
        if (line != null && session.plugin.routeRecorder != null) {
            session.plugin.routeRecorder.transferCart(line.id, previousCart, newCart)
        }
        if (session.state == TrainState.MOVING_BETWEEN_STATIONS) {
            movementAssistController.start()
        }
        session.debug("Transferred movement task to new minecart UUID=${newCart.uniqueId}")
    }

    fun setTeleporting(teleporting: Boolean) {
        session.isTeleporting = teleporting
    }

    fun canUsePortal(portalId: String): Boolean {
        val line = session.line
        return line != null && line.containsPortal(portalId)
    }

    fun scheduleSessionTask(task: Runnable, delay: Long, period: Long): Any? {
        val minecart = session.minecart ?: return null
        return trainScheduler.entityRun(minecart, task, delay, period)
    }

    fun getSession(): TrainSession = session

    @EventHandler(priority = EventPriority.NORMAL)
    fun onTrainEnterStop(event: TrainEnterStopEvent) {
        if (event.minecart != session.minecart) {
            return
        }

        if (!session.isTeleporting && !session.isPassengerStillRiding()) {
            handlePassengerExit()
            return
        }

        val enteredStop = event.stop
        if (enteredStop != null && session.targetStopId != null && session.targetStopId == enteredStop.id) {
            transitionToMovingInStation(enteredStop)
        }
    }

    @EventHandler(priority = EventPriority.HIGH)
    fun onVehicleMove(event: VehicleMoveEvent) {
        val minecart = session.minecart
        if (event.vehicle != minecart || minecart == null) {
            return
        }

        val line = session.line
        if (line != null) {
            session.plugin.routeRecorder.sample(line.id, minecart, event.to)
        }
        updateLastTravelDirection(event.from, event.to)

        if (session.state == TrainState.MOVING_BETWEEN_STATIONS) {
            val from = event.from
            val to = event.to
            if (from.world != null && to.world != null && from.world == to.world) {
                val dx = to.x - from.x
                val dz = to.z - from.z
                val distance = kotlin.math.sqrt(dx * dx + dz * dz)
                if (distance > 0.001) {
                    session.addDistance(distance)
                }
            }
        }

        if (session.state != TrainState.MOVING_IN_STATION) {
            return
        }

        val targetStop = session.plugin.stopManager.getStop(session.targetStopId)
        val targetStopLocation = targetStop?.stopPointLocation ?: return
        val currentLocation = minecart.location
        if (currentLocation.world == null || currentLocation.world != targetStopLocation.world) {
            return
        }

        val distance = currentLocation.distance(targetStopLocation)
        if (distance < 0.8) {
            transitionToStoppedAtStation(targetStop)
            return
        }

        physicsController.applyApproachBraking(minecart, distance, session.plugin.configFacade.getCartSpeed())
    }

    private fun transitionToStoppedAtStation(stop: Stop) {
        val minecart = session.minecart ?: return
        minecart.velocity = Vector(0, 0, 0)
        minecart.maxSpeed = 0.0
        movementAssistController.stop()

        val baseLocation = stop.stopPointLocation ?: return
        val snapLocation = baseLocation.clone()
        snapLocation.x = snapLocation.blockX + 0.5
        snapLocation.z = snapLocation.blockZ + 0.5
        val line = session.line
        if (line != null) {
            session.plugin.routeRecorder.sample(line.id, minecart, snapLocation)
        }
        // Disabled to avoid mounted minecart teleport desync for Bedrock players through Geyser.
        // SchedulerUtil.teleportEntity(minecart, snapLocation)

        settleDistanceFare(stop)

        val previousState = stateMachine.transitionTo(TrainState.STOPPED_AT_STATION, null)
        if (previousState == TrainState.MOVING_IN_STATION) {
            handleArrivalAtStation()
        }

        scoreboardController.updateBasedOnState(session)
    }

    private fun settleDistanceFare(stop: Stop) {
        val line = session.line ?: return
        val rule = line.priceRule ?: return
        val distance = session.distanceTraveled
        if (distance <= 0 && rule.getMode() != PriceRule.PricingMode.INTERVAL) {
            return
        }

        val passenger = session.passenger
        if (passenger == null || !passenger.isOnline) {
            return
        }

        val variablePrice =
            when (rule.getMode()) {
                PriceRule.PricingMode.DISTANCE -> distance * rule.getPerBlockRate()
                PriceRule.PricingMode.INTERVAL -> {
                    val intervals = session.plugin.priceService.countStopIntervals(line, session.entryStopId, stop.id)
                    intervals * rule.getPerIntervalRate()
                }

                else -> return
            }

        if (variablePrice > 0) {
            val status = session.plugin.ticketService.chargePrice(passenger, line, variablePrice)
            if (status == TicketService.TicketChargeStatus.CHARGED) {
                passenger.sendMessage(
                    session.plugin.languageManager.getMessage(
                        "economy.paid_distance",
                        LanguageManager.put(
                            LanguageManager.args(),
                            "price",
                            session.plugin.ticketService.format(variablePrice),
                        ),
                    ),
                )
            }
        }

        session.addDistance(-distance)
    }

    private fun transitionToMovingInStation(targetStop: Stop) {
        movementAssistController.stop()

        val previousState =
            stateMachine.transitionTo(
                TrainState.MOVING_IN_STATION,
                "enteredStop=${targetStop.id}",
            )
        if (previousState == TrainState.MOVING_BETWEEN_STATIONS) {
            eventPublisher.publishEnteringStop(targetStop)
        }

        scoreboardController.updateBasedOnState(session)
    }

    private fun transitionToMovingBetweenStations() {
        stateMachine.transitionTo(TrainState.MOVING_BETWEEN_STATIONS, null)

        val passenger = session.passenger
        if (passenger != null && passenger.isOnline) {
            passenger.sendTitle("", "", 0, 0, 0)
        }
        scoreboardController.updateBasedOnState(session)
    }

    private fun handleArrivalAtStation() {
        handleArrivalAtStation(false)
    }

    fun startAtStation() {
        handleArrivalAtStation(true)
    }

    private fun handleArrivalAtStation(isNewlySpawned: Boolean) {
        val line = session.line
        if (line == null) {
            cancel()
            return
        }

        if (!isNewlySpawned) {
            session.currentStopId = session.targetStopId
        }

        val currentStop = session.plugin.stopManager.getStop(session.currentStopId)
        if (currentStop == null) {
            session.debug("Cancelling train because current stop is missing: ${session.currentStopId}")
            cancel()
            return
        }

        if (line.getNextStopId(session.currentStopId) == null) {
            if (!handleTerminalStation()) {
                return
            }
            session.targetStopId = null
            eventPublisher.publishDockedAtStop(currentStop, true)
            return
        }

        eventPublisher.publishDockedAtStop(currentStop, false)
        scoreboardController.updateBasedOnState(session)
        scheduleNextDeparture()
    }

    private fun handleDeparture() {
        val line = session.line
        if (line == null) {
            cancel()
            return
        }

        session.refreshTargetFromCurrentStop()
        if (session.targetStopId == null) {
            return
        }

        val currentStop = session.plugin.stopManager.getStop(session.currentStopId)
        val nextStop = session.plugin.stopManager.getStop(session.targetStopId)
        if (currentStop == null || nextStop == null || currentStop.stopPointLocation == null) {
            session.debug(
                "Cancelling train because departure stops are incomplete: current=${session.currentStopId}, " +
                    "target=${session.targetStopId}",
            )
            cancel()
            return
        }

        val minecart = session.minecart ?: return
        session.plugin.routeRecorder.sample(line.id, minecart, currentStop.stopPointLocation)
        eventPublisher.publishDeparture(currentStop, nextStop)

        var maxSpeed = line.getMaxSpeed() ?: -1.0
        if (maxSpeed == -1.0) {
            maxSpeed = session.plugin.configFacade.getCartSpeed()
        }
        minecart.maxSpeed = maxSpeed

        val launchDirection = physicsController.initMinecartVelocity(minecart, currentStop.launchYaw)
        if (launchDirection != null) {
            session.lastTravelDirection = launchDirection
        }

        transitionToMovingBetweenStations()
        movementAssistController.start()
    }

    private fun scheduleNextDeparture() {
        val delay = session.plugin.configFacade.getCartDepartureDelay()
        session.debug(
            "Schedule departure in $delay ticks for passenger=${session.safePassengerName()}, " +
                "currentStop=${session.currentStopId}",
        )
        val minecart = session.minecart ?: return
        trainScheduler.entityRun(
            minecart,
            Runnable {
                if (session.isPassengerStillRiding()) {
                    handleDeparture()
                } else {
                    handlePassengerExit()
                }
            },
            delay,
            -1,
        )
    }

    private fun handleTerminalStation(): Boolean {
        val passenger = session.passenger
        val line = session.line
        if (passenger == null || !passenger.isOnline || line == null) {
            cancel()
            return false
        }
        session.debug(
            "Terminal station reached for passenger=${session.safePassengerName()}, stop=${session.currentStopId}",
        )

        val routeResult = session.plugin.routeRecorder.finishIfRecording(line.id, session.minecart)
        notifyRouteRecorder(routeResult)
        if (routeResult.status == org.cubexmc.metro.manager.RouteRecorder.FinishResult.Status.TOO_FEW_POINTS) {
            session.plugin.logger.warning(
                "[RouteRecorder] Route recording for line ${line.id} reached the terminal but only collected " +
                    "${routeResult.pointCount} point(s).",
            )
        }

        val minecart = session.minecart
        if (minecart == null) {
            cancel()
            return false
        }

        trainScheduler.entityRun(
            minecart,
            Runnable {
                val currentMinecart = session.minecart
                if (currentMinecart != null && !currentMinecart.isDead) {
                    currentMinecart.eject()
                    session.plugin.scoreboardManager.clearPlayerDisplay(passenger)

                    trainScheduler.entityRun(
                        currentMinecart,
                        Runnable {
                            if (!currentMinecart.isDead) {
                                currentMinecart.remove()
                            }
                            cancel()
                        },
                        40L,
                        -1,
                    )
                } else {
                    cancel()
                }
            },
            60L,
            -1,
        )
        return true
    }

    private fun notifyRouteRecorder(result: org.cubexmc.metro.manager.RouteRecorder.FinishResult) {
        val recorderId = result.recorderId ?: return
        if (result.status == org.cubexmc.metro.manager.RouteRecorder.FinishResult.Status.NOT_RECORDING) {
            return
        }
        val recorder = Bukkit.getPlayer(recorderId)
        if (recorder == null || !recorder.isOnline) {
            return
        }

        val args = LanguageManager.args()
        LanguageManager.put(args, "line_id", result.lineId)
        LanguageManager.put(args, "point_count", result.pointCount.toString())

        val key =
            when (result.status) {
                org.cubexmc.metro.manager.RouteRecorder.FinishResult.Status.SAVED -> "line.record_saved"
                org.cubexmc.metro.manager.RouteRecorder.FinishResult.Status.TOO_FEW_POINTS -> "line.record_too_few"
                org.cubexmc.metro.manager.RouteRecorder.FinishResult.Status.FAILED -> "line.record_failed"
                org.cubexmc.metro.manager.RouteRecorder.FinishResult.Status.NOT_RECORDING -> null
            }
        if (key != null) {
            recorder.sendMessage(session.plugin.languageManager.getMessage(key, args))
        }
        if (result.status == org.cubexmc.metro.manager.RouteRecorder.FinishResult.Status.TOO_FEW_POINTS) {
            recorder.sendMessage(session.plugin.languageManager.getMessage("line.record_too_few_hint"))
        }
    }

    private fun handlePassengerExit() {
        cancel()
    }

    private fun updateLastTravelDirection(from: Location?, to: Location?) {
        if (session.state == TrainState.STOPPED_AT_STATION || from == null || to == null) {
            return
        }
        if (from.world == null || to.world == null || from.world != to.world) {
            return
        }
        val direction = to.toVector().subtract(from.toVector())
        if (direction.lengthSquared() < 0.0001) {
            return
        }
        session.lastTravelDirection = direction.normalize()
    }

    companion object {
        @JvmStatic
        fun getTaskFor(cart: Minecart?): TrainMovementTask? = TrainTaskRegistry.get(cart)

        @JvmStatic
        fun shutdownActiveTasks(): Int = TrainTaskRegistry.shutdownActiveTasks()

        @JvmStatic
        fun shutdownActiveTasks(plugin: Metro, folia: Boolean): Int =
            TrainTaskRegistry.shutdownActiveTasks(plugin, folia)

        @JvmStatic
        fun startTrainTask(
            plugin: Metro,
            minecart: Minecart,
            passenger: Player?,
            lineId: String,
            currentStopId: String,
        ) {
            TrainTaskStarter.start(plugin, minecart, passenger, lineId, currentStopId)
        }
    }
}
